### PR TITLE
refactor: improve `PrincipalInner` encapsulation

### PR DESCRIPTION
**Pull request by Daniel-Bloom-dfinity**

### CODE REVIEW

File: ic-types/src/principal.rs [hunk 1]
```diff
@@ -185,9 +483,87 @@ impl Principal {
...
+        let mut buffer = Buffer::new();
+        for b in bytes {
+            if buffer.try_push(b).is_some() {
+                return Err(PrincipalBufferError::BufferTooLong());
+            }
+        }
+        Principal::try_from_buffer(buffer)
+    }
+
+    /// Parse raw binary into a Principal. `panic`s if the bytes can't be interpretted.
```
Comment (ggreif) [inline ic-types/src/principal.rs:499]:
```suggestion
    /// Parse raw binary into a Principal. `panic`s if the bytes can't be interpreted.
```

File: ic-types/src/principal.rs [hunk 2]
```diff
@@ -17,11 +199,20 @@ pub enum PrincipalError {
...
-const ID_ANONYMOUS_BYTES: &[u8] = &[PrincipalClass::Anonymous as u8];
+impl From<PrincipalBufferError> for PrincipalError {
+    fn from(v: PrincipalBufferError) -> Self {
+        match v {
+            PrincipalBufferError::BufferTooLong() => Self::BufferTooLong(),
+        }
+    }
+}
 
 /// A class of principal. Because this should not be exposed it
```
Comment (ggreif) [inline ic-types/src/principal.rs:217]: not your PR, but this sentence abruptly

**Comment by Daniel-Bloom-dfinity** [comment]
Repurposing this PR for doing preliminary refactors instead.

**Review by hansl — APPROVED** [review]
LGTM. Good work!

**Pull request merged**